File: _03_AlgorithmsOnGraphs/Java/src/_05_Ex2_DeterminingOrderOfCourses/Toposort.java
package _05_Ex2_DeterminingOrderOfCourses; // must be removed for submission

import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;


public class Toposort // DECOMPOSITION OF GRAPHS 2: EXCERCISE 2 - Compute a topological ordering of a given directed acyclic graph (DAG) with n vertices and m edges.
{
	private static ArrayList<Integer> toposort(ArrayList<Integer>[] adj) // we are going to put the elements on a STACK, as we start placing the elements from the end to the start
	{
		boolean[] nodes = new boolean[adj.length]; // the storage for node status has to be initialized with unvisited values
		ArrayList<Integer> order = new ArrayList<Integer>(); // this is for storing the TOPOLOGICAL SORT

		for (int v = 0; v < nodes.length; v++)
		{
			if (!nodes[v])
				dfs(adj, nodes, order, v);
		}

		Collections.reverse(order);
		return order;
	}
				
    private static void dfs(ArrayList<Integer>[] adj, boolean[] nodes, ArrayList<Integer> order, int v) // helper function that will look around the given VERTEX and find all other connected VERTICES (direct and indirect)
	{
    	nodes[v] = true;

		for (int w : adj[v])
		{
			if (!nodes[w])
				dfs(adj, nodes, order, w);
		}

		order.add(v); // this is where we normally store the 'post-order' value for topological sort - where we exit the recursion
	}

    
    public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
        }
        
        ArrayList<Integer> order = toposort(adj); // Good job! (Max time used: 2.17/3.00, max memory used: 207482880/536870912.)
        
        for (int x : order) {
            System.out.print((x + 1) + " ");
        }
    }
}

File: _03_AlgorithmsOnGraphs/CSharp/_05_Ex2_DeterminingOrderOfCourses/Toposort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace _05_Ex2_DeterminingOrderOfCourses // DECOMPOSITION OF GRAPHS 2: EXCERCISE 2 - Compute a topological ordering of a given directed acyclic graph (DAG) with 𝑛 vertices and 𝑚 edges.
{
    class TopologicalSort // the whole solution is based on STATIC method 'Toposort' as such was the provided Java template.
    {
        static Stack<int> Toposort(List<int>[] adj) // we are going to put the elements on a STACK, as we start placing the elements from the end to the start
        {
            bool[] nodes = new bool[adj.Length]; // the storage for node status has to be initialized with unvisited values
            Stack<int> order = new Stack<int>(); // this is for storing the TOPOLOGICAL SORT

            for (int v = 0; v < nodes.Length; v++)
            {
                if (!nodes[v])
                    DFS(adj, nodes, order, v);
            }

            return order;
        }


        static void DFS(List<int>[] adj, bool[] nodes, Stack<int> order, int v) // helper function that will look around the given VERTEX and find all other connected VERTICES (direct and indirect)
        {
            nodes[v] = true;

            foreach (int w in adj[v])
                if (!nodes[w])
                    DFS(adj, nodes, order, w);

            order.Push(v); // this is where we normally store the 'post-order' value for topological sort - where we exit the recursion
        }


        static void Main(string[] args)
        {
            var input = Console.ReadLine().Split();
            int n = int.Parse(input[0]);
            int m = int.Parse(input[1]);

            List<int>[] adj = new List<int>[n];
            for (int i = 0; i < n; i++)
                adj[i] = new List<int>();

            for (int i = 0; i < m; i++)
            {
                int x, y;
                input = Console.ReadLine().Split();
                x = int.Parse(input[0]);
                y = int.Parse(input[1]);
                adj[x - 1].Add(y - 1); // in ORDERED GRAPHS we take the start vertex and append the END VERTEX to it, applying correction for 0-based indexing
            }

            Stack<int> order = Toposort(adj); // Good job! (Max time used: 0.54/3.00, max memory used: 37621760/536870912.)
            foreach (int x in order)
                Console.Write((x + 1) + " ");

            Console.ReadKey();
        }
    }
}
